File: demo3_books.py
#!/usr/bin/env python3
"""
Books.toscrape 电商数据采集 - 演示 Playwright 动态渲染能力
输出: JSON + CSV
"""
import json
import csv
import asyncio
from datetime import datetime

async def scrape_books():
    """用 Playwright 采集书籍数据（JS渲染页面）"""
    try:
        from playwright.async_api import async_playwright
    except ImportError:
        print("Playwright not installed, using requests fallback")
        return await scrape_books_requests()
    
    books = []
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()
        
        # 采集前3页
        for page_num in range(1, 4):
            url = f"http://books.toscrape.com/catalogue/category/books_1/page-{page_num}.html"
            print(f"  采集: {url}")
            await page.goto(url, wait_until="networkidle")
            
            # 等待书籍卡片加载
            await page.wait_for_selector("article.product_pod", timeout=10)
            
            # 提取数据
            items = await page.query_selector_all("article.product_pod")
            for item in items:
                title_elem = await item.query_selector("h3 a")
                price_elem = await item.query_selector(".price_color")
                rating_elem = await item.query_selector(".star-rating")
                img_elem = await item.query_selector("img")
                
                title = await title_elem.get_attribute("title") if title_elem else ""
                price = await price_elem.inner_text() if price_elem else ""
                rating = await rating_elem.get_attribute("class") if rating_elem else ""
                img = await img_elem.get_attribute("src") if img_elem else ""
                
                books.append({
                    "title": title,
                    "price": price.strip(),
                    "rating": rating.replace("star-rating ", "").strip(),
                    "image": "http://books.toscrape.com/" + img if img else "",
                    "page": page_num,
                    "scraped_at": datetime.now().strftime("%Y-%m-%d %H:%M"),
                })
        
        await browser.close()
    
    return books

async def scrape_books_requests():
    """Fallback: 用 requests + BeautifulSoup"""
    import requests
    from bs4 import BeautifulSoup
    
    books = []
    headers = {"User-Agent": "Mozilla/5.0 (compatible; GaryPortfolio/1.0)"}
    
    for page_num in range(1, 4):
        url = f"http://books.toscrape.com/catalogue/category/books_1/page-{page_num}.html"
        print(f"  采集: {url}")
        resp = requests.get(url, headers=headers, timeout=15)
        soup = BeautifulSoup(resp.text, "html.parser")
        
        items = soup.select("article.product_pod")
        for item in items:
            title = item.select_one("h3 a")["title"]
            price = item.select_one(".price_color").text.strip()
            rating = item.select_one(".star-rating")["class"]
            books.append({
                "title": title,
                "price": price,
                "rating": " ".join(rating),
                "page": page_num,
            })
    
    return books

def save_json(data, path):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

def save_csv(data, path):
    if not data:
        return
    with open(path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=data[0].keys())
        writer.writeheader()
        writer.writerows(data)

if __name__ == "__main__":
    print("=== 电商数据采集 (Playwright 动态渲染) ===")
    books = asyncio.run(scrape_books())
    
    print(f"\n共采集 {len(books)} 本书:")
    for b in books[:8]:
        print(f"  [{b['rating']}] {b['title'][:50]} - {b['price']}")
    
    save_json(books, "/home/peachy/.openclaw/workspace/portfolio/demo3_books.json")
    save_csv(books, "/home/peachy/.openclaw/workspace/portfolio/demo3_books.csv")
    print(f"\n数据已保存: demo3_books.json + demo3_books.csv")

File: demo3_books.mjs
#!/usr/bin/env node
/**
 * 电商数据采集 - Playwright 动态渲染演示
 * 目标: books.toscrape.com (专门用于爬虫练习的安全站点)
 */
import { chromium } from '/home/peachy/.openclaw/workspace/node_modules/playwright/index.mjs';
import { writeFileSync } from 'fs';

const browser = await chromium.launch({ headless: true });
const page = await browser.newPage();

const allBooks = [];

console.log('=== 电商数据采集 (Playwright 动态渲染) ===');

for (let pageNum = 1; pageNum <= 3; pageNum++) {
    const url = `http://books.toscrape.com/catalogue/category/books_1/page-${pageNum}.html`;
    console.log(`  采集: ${url}`);
    
    await page.goto(url, { waitUntil: 'networkidle', timeout: 15000 });
    
    // 等待书籍卡片
    await page.waitForSelector('article.product_pod', { timeout: 10000 }).catch(() => {});
    
    // 提取书籍数据
    const books = await page.evaluate(() => {
        const items = document.querySelectorAll('article.product_pod');
        return Array.from(items).map(item => {
            const title = item.querySelector('h3 a')?.getAttribute('title') || '';
            const price = item.querySelector('.price_color')?.textContent?.trim() || '';
            const rating = item.querySelector('.star-rating')?.className?.replace('star-rating', '').trim() || '';
            const img = item.querySelector('img')?.getAttribute('src') || '';
            return { title, price, rating: rating.replace('star-rating', '').trim(), image: img };
        });
    });
    
    console.log(`    获取 ${books.length} 本书`);
    allBooks.push(...books.map(b => ({ ...b, page: pageNum })));
}

await browser.close();

console.log(`\n共采集 ${allBooks.length} 本书`);
console.log('\n示例数据:');
allBooks.slice(0, 6).forEach(b => {
    console.log(`  [${b.rating}] ${b.title.substring(0, 50)} - ${b.price}`);
});

// 保存 JSON
writeFileSync('/home/peachy/.openclaw/workspace/portfolio/demo3_books.json', 
    JSON.stringify({ source: 'books.toscrape.com', total: allBooks.length, books: allBooks }, null, 2));
console.log('\n数据已保存: demo3_books.json');
